File: medium/1664-ways-to-make-a-fair-array/solution.py
from typing import List


class Solution:
    def waysToMakeFair(self, nums: List[int]) -> int:
        """
        1664. Ways to Make a Fair Array
        Time: O(n)
        Space: O(1)
        """
        n = len(nums)
        total_even = sum(nums[i] for i in range(0, n, 2))
        total_odd = sum(nums[i] for i in range(1, n, 2))

        pre_even = pre_odd = 0
        count = 0
        for i, x in enumerate(nums):
            # suffix sums (after i), excluding nums[i] itself
            if i % 2 == 0:
                suf_even = total_even - pre_even - x
                suf_odd = total_odd - pre_odd
            else:
                suf_even = total_even - pre_even
                suf_odd = total_odd - pre_odd - x
            # removal flips suffix parity
            if pre_even + suf_odd == pre_odd + suf_even:
                count += 1
            # advance prefix with nums[i]
            if i % 2 == 0:
                pre_even += x
            else:
                pre_odd += x
        return count

File: medium/1664-ways-to-make-a-fair-array/solution.java
/**
 * 1664. Ways to Make a Fair Array
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int waysToMakeFair(int[] nums) {
        int n = nums.length;
        long totalEven = 0, totalOdd = 0;
        for (int i = 0; i < n; i++) {
            if (i % 2 == 0) {
                totalEven += nums[i];
            } else {
                totalOdd += nums[i];
            }
        }

        long preEven = 0, preOdd = 0;
        int count = 0;
        for (int i = 0; i < n; i++) {
            int x = nums[i];
            long sufEven, sufOdd;
            if (i % 2 == 0) {
                sufEven = totalEven - preEven - x;
                sufOdd = totalOdd - preOdd;
            } else {
                sufEven = totalEven - preEven;
                sufOdd = totalOdd - preOdd - x;
            }
            // removal flips suffix parity
            if (preEven + sufOdd == preOdd + sufEven) {
                count++;
            }
            if (i % 2 == 0) {
                preEven += x;
            } else {
                preOdd += x;
            }
        }
        return count;
    }
}
